### ISSUE TITLE
Account for tilde expansion in [[ expression ]]

### ISSUE THREAD

**Issue by lollipopman** [koalaman/shellcheck#1769]
#### For bugs
- Rule Id: SC2193
- My shellcheck version: 0.7.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```bash
#!/bin/bash

if [[ ~ == ~+ ]]; then
  echo 'In Home'
fi
```

#### Here's what shellcheck currently says:

```
In test.sh line 3:
if [[ ~ == ~+ ]]; then
        ^-- SC2193: The arguments to this comparison can never be equal. Make sure your syntax is correct.
```

#### Here's what I wanted or expected to see:

Nothing, since with tilde expansion they can be equal

**Comment by koalaman** [comment]
The support for tilde expansion in this and other contexts has been significantly improved. Thanks!

**Comment by lollipopman** [comment]
awesome thanks for all your work on shellcheck!

**Comment by jnpkrn** [comment]
Indeed, thanks a lot.

Sorry for this excatation, I am afraid that wider theme of _tilde
expansion_ on unexpected places, notably within `[[ expr ]]` deserves
more attention/checking/hinting, since its semantics being plugged in
is quite unexpected to many (incl. me until recently).

It would be really nice if something like `[[ "$1" = ~* ]]` was brought
into one's attention with a message like:

```
~* will try to expand to the home directory of user * (star) rather
than being used as a pattern match operand
```

Case in point: https://github.com/rpm-software-management/dnf/pull/1730

Sorry for a message not too in-line with `ShellCheck`'s style, but
you get the point.  Also, let me know if a separate issue is warranted;
jumped here for the very combination of the two parts of the Bash
language.
